File: backend/app/schemas/budget.py
from pydantic import BaseModel
from decimal import Decimal


class BudgetAssignmentUpsert(BaseModel):
    category_id: str
    month: str  # YYYY-MM
    assigned_amount: Decimal


class BudgetAssignmentResponse(BaseModel):
    id: str
    household_id: str
    category_id: str
    month: str
    assigned_amount: Decimal

    model_config = {"from_attributes": True}


class CategoryBudgetRow(BaseModel):
    category_id: str
    category_name: str
    group_id: str
    assigned: Decimal
    activity: Decimal
    available: Decimal
    carryover: Decimal


class GroupBudgetRow(BaseModel):
    group_id: str
    group_name: str
    sort_order: int
    is_income: bool
    assigned: Decimal
    activity: Decimal
    available: Decimal
    carryover: Decimal
    categories: list[CategoryBudgetRow]


class BudgetMonthResponse(BaseModel):
    month: str
    total_income: Decimal
    total_assigned: Decimal
    total_activity: Decimal
    total_available: Decimal
    # Envelope rollover (docs/superpowers/specs/2026-07-02-budget-rollover-design.md):
    # `available` is cumulative (carryover + assigned + activity).
    ready_to_assign: Decimal
    total_carryover_in: Decimal
    overspend_deducted: Decimal
    groups: list[GroupBudgetRow]

File: frontend/src/lib/api/budget.ts
import api from "./client";

export interface CategoryBudgetRow {
  category_id: string;
  category_name: string;
  group_id: string;
  assigned: number;
  activity: number;
  available: number;
  carryover: number;
}

export interface GroupBudgetRow {
  group_id: string;
  group_name: string;
  sort_order: number;
  is_income: boolean;
  assigned: number;
  activity: number;
  available: number;
  carryover: number;
  categories: CategoryBudgetRow[];
}

export interface BudgetMonthResponse {
  month: string;
  total_income: number;
  total_assigned: number;
  total_activity: number;
  total_available: number;
  ready_to_assign: number;
  total_carryover_in: number;
  overspend_deducted: number;
  groups: GroupBudgetRow[];
}

export interface BudgetAssignment {
  id: string;
  household_id: string;
  category_id: string;
  month: string;
  assigned_amount: number;
}

export const budgetApi = {
  getMonth: (month: string) =>
    api.get<BudgetMonthResponse>(`/budget/month/${month}`).then((r) => r.data),

  assign: (data: { category_id: string; month: string; assigned_amount: number }) =>
    api.put<BudgetAssignment>("/budget/assign", data).then((r) => r.data),

  copyMonth: (source_month: string, target_month: string) =>
    api.post<{ copied: number }>("/budget/copy-month", { source_month, target_month }).then((r) => r.data),
};
